File: 890-find-and-replace-pattern/890-find-and-replace-pattern.py
class Solution(object):
    
    def findAndReplacePattern(self, words, pattern):
        """
        :type words: List[str]
        :type pattern: str
        :rtype: List[str]
        """
        # Solution: O(n m) time - n is num words and m is length of words - O(n) additional space.
        
        output = []
        
        for word in words:
            if self.check(word, pattern):
                output.append(word)
                
        return output
    
    
    def check(self, word, pattern):
        m1, m2 = {}, {}

        for a, b in zip(word, pattern): # Check over characters to verify bijection.
            if a not in m1: 
                m1[a] = b
            if b not in m2: 
                m2[b] = a

            if a != m2[b] or b != m1[a]:
                return False

        return True
        
        

File: 890-find-and-replace-pattern/890-find-and-replace-pattern.java
class Solution {
    
    public List<String> findAndReplacePattern(String[] words, String pattern) {
        
        // Solution: O(n m) time - n is num words and m is length of words - O(n) additional space.
        // Note: Java translation of my Python solution.
        
        List<String> output = new ArrayList<String>();
        
        for (String word: words) if (this.check(word, pattern)) output.add(word);
        
        return output;
            
    }
    
    public boolean check(String word, String pattern) {
        
        HashMap<Character, Character> m1 = new HashMap<>();
        HashMap<Character, Character> m2 = new HashMap<>();
        
        for (int i = 0; i < word.length(); i++) {
            
            char a = word.charAt(i);
            char b = pattern.charAt(i);
            
            if (!m1.containsKey(a)) m1.put(a, b);
            if (!m2.containsKey(b)) m2.put(b, a);
                
            if (a != m2.get(b) || b != m1.get(a)) return false;
            
        }
        
        return true;
        
    }

    
}